File: backend/schemas/task.py
# schemas/task.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional, Dict, List
from schemas.tag import Tag

class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    completed: bool = False
    deadline: Optional[datetime] = None

class TaskCreate(TaskBase):
    tags: Optional[List[int]] = []

class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    completed: Optional[bool] = None
    deadline: Optional[datetime] = None
    tags: Optional[List[int]] = None

class Task(TaskBase):
    id : int
    user_id :  int
    created_at : datetime
    updated_at : datetime
    tags : List[Tag] = []

    model_config = ConfigDict(from_attributes=True)

class TaskSummary(BaseModel):
    total_completed: int
    total_pending: int
    by_tag: Dict[str, int]

File: frontend/types/task.ts
import { Tag } from './tag';

export interface Task {
  id: number;
  title: string;
  description?: string;
  completed: boolean;
  deadline?: string | null;
  tags: Tag[];
}

export interface TaskCreate {
  title: string;
  description?: string;
  deadline?: string | null;
  tags?: number[];
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  completed?: boolean;
  deadline?: string | null;
  tags?: number[];
}

export interface TaskSummary {
  total: number;
  total_completed: number;
  total_pending: number;
  by_tag: Record<string, number>;
}